[AArch64] Optimize vector fmul(sitofp/uitofp, 1/2^N) -> scvtf/ucvtf

jph-13 opened the pull request:
When a vector integer-to-float conversion is followed by a multiply with a
  reciprocal power-of-two constant, we can fold both operations into a single
  SCVTF or UCVTF instruction with a fixed-point shift operand.

  For example, `fmul(sitofp(v2i32 x), <0.5, 0.5>)` becomes `scvtf.2s v0, v0, #1`.

  This is a reworked version with several improvements over the original submission:
  - Rewrite the C++ operand matcher to share implementation with the existing
    `SelectCVTFixedPointVec` (MOVIshift, FMOV, and DUP handling with correct
    truncation for f16)
  - Add `uitofp`/`ucvtf` patterns via a `CVTFRecipPat` multiclass
  - Add full GlobalISel support (`GIComplexOperandMatcher` + renderer)

  Supported vector types: `v2f32`, `v4f32`, `v2f64`, `v4f16`, `v8f16`.

  Fixes #94909

stephentyrone commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
+
+  // Check if IntrinsicC is within the valid range [1, FloatBits].
+  // The 's' value must be in [1, FloatBits].
+  if (IntrinsicC <= 0 || IntrinsicC > FloatBits)
```
This check makes the preceding check entirely redundant; I think that's OK, if you just want to be super explicit.

stephentyrone commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
+
+  MVT FloatTy = N->getSimpleValueType(0).getVectorElementType();
+  int32_t FloatBits = FloatTy.getSizeInBits();
+  if (FloatBits != 32 && FloatBits != 64)
```
IIRC scvtf supports half-precision as well.

stephentyrone commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
+    return SDValue();
+
+  if (IntBits > FloatBits)
+    return SDValue();
```
Maybe worth a comment here noting that we cannot just convert to a wider float type and then narrow because it would introduce a potential double-rounding.

stephentyrone commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
+    return -1;
+
+  // Get raw bits
+  APInt Bits = FirstVal.bitcastToAPInt();
```
Is there a reason why we do all the rest of this in terms of bit patterns instead of doing `frexp` on the APFloat value to extract the exponent and significand and working with those?

stephentyrone commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
+
+/// Fold a floating-point multiply by power of two into fixed-point to
+/// floating-point conversion.
+static SDValue performFMulCombine(SDNode *N, SelectionDAG &DAG,
```
Maybe worth a comment that this optimization would not be valid for a constrained FMul intrinsic, because sitofp rounds according to the default rounding mode (RN) but scvtf uses the rounding mode from FPCR.

jph-13 commented:
I will take a look at the tablegen impls. Maybe that will help me understand why I am having issues with half too.

jph-13 commented:
I just resolved all the original flags since the new implementation is very different. I did try to get f16 working but I became very confused. As of now it doesn't appear to have a match in TD. I started creating one but I am not sure if I shold replace all the round tripping or not. So I figured I would see if we could get this in, then maybe try another pass at half later.

jph-13 commented:
This is incomplete, but I could really use some help. I have never touched tablegen before this and feel I am making it way too difficult. 

I am having specific issues with the smaller registers and sizes (are my matchers too restrictive?). I left in commented out vNi16 code for I can't get any of it to not conflict. The v1i32 and v1i64 are also escaping me. As for the f16 implementations, it seems I need to handle a sext, but I figure the rest should be working first. 

I could use any comments or guidance folks have.

Thanks.

davemgreen reviewed (commented):
It looks like this got quite far. I've left a few comments inline, tablegen can be finicky at times.

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrFormats.td:
```diff
@@ -5407,6 +5433,102 @@ class BaseIntegerToFPUnscaled<bits<2> rmode, bits<3> opcode,
+//   (SCVTFv2f32_v2i64_V FPR128:$Rn, fixedpoint_recip_v2f32_v2i64:$scale)
+// >;
+
+def : Pat<(fmul(sint_to_fp(v2i64 FPR128:$Rn)),
```
That does look like it is being promoted. It doesn't happen here: https://godbolt.org/z/xdq14cT67. Any idea what your test might be doing differently?

jph-13 replied:
OK, the error was not with the test but rather with the launch command I had for debugging. I repaired that now, sorry for the silly error. 
```
// %bb.0:                               // %entry
        movi.4h v1, #52, lsl #8
        scvtf.4h        v0, v0
        fmul.4h v0, v0, v1
        ret
.Lfunc_end0:
```
Now I think it is close but I still fail to extract the constant for the f16 cases. Using the patterns above it does match. The op comes in as an `AArch64ISD::NVCAST` which then has a 0 operand that is an `AArch64ISD::MOVIshift`. So I seem to be too deep to figure out if it is a splat or not.

davemgreen reviewed (commented):
Sorry I missed the update. Nice job getting this working, it looks good. I just have a couple of nitpicks, it looks good to me otherwise.

jph-13 commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -3952,6 +3960,129 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
+                                                         bool isReciprocal) {
+
+  // N must be a bitcast/nvcast of a vector float type.
+  if (!((N.getOpcode() == ISD::BITCAST ||
+         N.getOpcode() == AArch64ISD::NVCAST) &&
+        N.getValueType().isVector() && N.getValueType().isFloatingPoint())) {
```
But the input is a vector of ints at this point no?

jph-13 commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -3952,6 +3960,129 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
+  // --- Perform fixed-point reciprocal check and power-of-2 validation on FVal
+  // --- Normalize f16 to f32 if needed for consistent APFloat operations.
+  if (N.getValueType().getVectorElementType() == MVT::f16) {
+    bool ignored;
```
This code was a relic from an AI conversation. I don't think the code was needed. Originally I was doing more intermediate  math. I removed the logic.

davemgreen reviewed (commented):
Thanks - I think the handling of MOVIshift might need to be tightened up a bit, otherwise it could ignore important certain bits of the constant. Detecting the constant correctly is the most awkward part of this with all the different combos, but luckily it looks like many of them use a similar node type.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -3952,6 +3960,132 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
+      // Unsupported floating-point element type.
+      return false;
+    }
+  } else if (auto *CFP =
+                 dyn_cast<ConstantFPSDNode>(ImmediateNode.getOperand(0))) {
```
Does this case come up?

jph-13 replied:
Not in the test cases I have. Another cautionary reiic from an AI conversation. I will drop it.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -3952,6 +3960,132 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
+                                                         unsigned FloatWidth,
+                                                         bool IsReciprocal) {
+
+  SDValue ImmediateNode;
+  // N must be a bitcast, nvcast, or fmov
+  if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST ||
+      N.getOpcode() == AArch64ISD::FMOV) {
+    ImmediateNode = N.getOperand(0);
```
I would suggest that this be a little tighter on what it considers, to prevent any (rare) cases from treating the node with an incorrect type (like looking at operand 0 where it doesn't exist and how FMOV is currently handled.

I would probably make it something like
```
SDValue ImmediateNode = N;
if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST)
  ImmediateNode = N.getOperand(0);
```

Then check the node type is still a vector (in case it is bitcast to a scalar) `ImmediateNode.getValueType().isVector()`. 
The if below would check `ImmediateNode.getOpcode() == AArch64ISD::FMOV`, not for ISD::Constant, and check its operand (which we know should always be a ConstantSDNode.

That should hopefully make it more resilient to odd cases that don't come up very often, and hopefully a little more clear which nodes equate to what.

davemgreen reviewed (commented):
Thanks for adding all the tests, that's fantastic. I have a suggestion to make the code slightly more structured in case something happens with weird node types, but it LGTM otherwise.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -3952,6 +3960,125 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
+                ImmediateNode.getConstantOperandVal(0)
+                    << ImmediateNode.getConstantOperandVal(1));
+    IsIntConstant = true;
+  } else if (ImmediateNode.getOpcode() == ISD::Constant) {
```
Could we end up with:
 - N is some random node with 1 constant operand. Say a MOVImsl.
 - Nothing checks N's opcode, only that is has >= 1 operand.
 - The ImmediateNode turns out to be a ISD::Constant, which is treated as a FMOV operand even though it wasn't.

(Granted a MOVImsl will no naturally be a fp type but it could be some other random operand type to the fmul).

jph-13 replied:
Fair point. I fixed my shielding code and removed the FMUL from above.

```
  if (!(ImmediateNode.getOpcode() == AArch64ISD::DUP ||
        ImmediateNode.getOpcode() == AArch64ISD::MOVIshift ||
        ImmediateNode.getOpcode() == ISD::BUILD_VECTOR ||
        ImmediateNode.getOpcode() == ISD::Constant ||
        ImmediateNode.getOpcode() == ISD::SPLAT_VECTOR)) {
    return false;
  }

  if (ImmediateNode.getOpcode() != ISD::Constant) {
    auto *C = dyn_cast<ConstantSDNode>(ImmediateNode.getOperand(0));
    if (!C)
      return false;
  }
```
  
The ISD::Constant logic is needed for ```test_v4f32_div_const_4```

jph-13 committed:
[AArch64] Optimize vector fmul(sitofp/uitofp(x), 1/2^N) to scvtf/ucvtf with shift

When a vector integer-to-float conversion is followed by a multiply with a
reciprocal power-of-two constant, we can fold both operations into a single
SCVTF or UCVTF instruction with a fixed-point shift operand.

For example, fmul(sitofp(v2i32 x), <0.5, 0.5>) becomes scvtf.2s v0, v0, #1.

This is a reworked version of #141480 with several improvements:
- Rewrite the C++ operand matcher to share implementation with the existing
  SelectCVTFixedPointVec (MOVIshift, FMOV, and DUP handling with correct
  truncation for f16)
- Add uitofp/ucvtf patterns via a CVTFRecipPat multiclass
- Add full GlobalISel support (GIComplexOperandMatcher + renderer)

Supported vector types: v2f32, v4f32, v2f64, v4f16, v8f16.

Fixes #94909

jph-13 commented:
Sorry, this had fell in the priority list. I have some extra cycles now, hoping to get accross the line.

davemgreen reviewed (approved):
LGTM - thanks for doing this, it looks great.

Do you have commit access or are you happy for me to merge?

jph-13 commented:
@davemgreen, I do not have commit. Could you please do it? Thank you in advance.
